Create React Appの提出物

ham-cap opened the pull request:
Reactの課題でメモアプリを作成いたしました。
よろしくお願いいたします。

ham-cap committed:
不要なファイルを削除・App.jsとその他のコンポーネントの動作確認

ham-cap committed:
ESLintとPrettierを設定

ham-cap committed:
FormSwitcherによるCreateFormとEditFormの切り替えを実装

ham-cap committed:
FormSwitcherの親要素としてMemoListを作成した

・一覧表示された既存のメモと下段の＋マークのどちらがクリックされたかで表示されるフォームを制御するため、めもの一覧表示と＋マークの表示を担当するMemoListをFormSwitcherの親要素として設定した。

ham-cap committed:
CreateFormのinput要素とApp.jsのstateをバインディング

ham-cap committed:
prop-typesをインストール

・propsにバリデーションをかけるためにprop-typesをインストールした。

ham-cap committed:
prop-typesを用いてpropsの型チェックを追加

ham-cap committed:
不要なコメントを削除・CreateFormのinputをtextareaに変更

ham-cap committed:
新規登録機能と一覧表示機能を実装

ham-cap committed:
新規登録処理を修正・一覧表示を修正

・1行目をタイトルとして使用するため、新規登録時に文字列を改行文字で区切り配列として保存されるようにした。
そのうえで、一覧表示の際に各要素の1行目のみを出力するよう変更した。

ham-cap committed:
新規登録用フォームの表示切り替えを実装

・＋マークをクリックすると新規登録用フォームが表示できる必要があったため実装した。

ham-cap committed:
編集用フォームの見た目と表示切り替えを実装

・一覧で表示されている既存のメモをクリックすると編集フォームが表示される必要があるため実装した。

ham-cap committed:
編集用フォームが呼び出された際、選択されたメモの本文が初期値として表示されるようにした

ham-cap committed:
EditForm.cssを追加

・textareaタグのデフォルトの大きさでは表示が小さすぎ、調整する必要があったため。

ham-cap committed:
編集機能を実装

ham-cap committed:
削除機能を実装・indexOfSelectedMemoの初期値をnullに変更

・MemoListのstateであるindexOfSelectedMemoの初期値を空のStringにしており、indexが渡ってきたときだけNumberの値が入るようになっていたが、indexが渡って来なかった場合に型チェックで引っかかってしまうため、初期値をnullに変更した。

ham-cap committed:
新規登録、更新、削除の各処理の最後でフォームを非表示にするよう修正

・各処理の終了後もフォームが表示されていると、誤操作の原因になりそうだったため、処理の終了時に必ず非表示に切り替えることとした。

ham-cap committed:
既存メモの取得タイミングを変更

・localStorageに保存されている既存データはTodoAppのconstructorでstateを定義する直前にフェッチしていたが、このタイミングではstateの定義が完了している保証
がないため、確実にstateの準備ができているマウント直後にフェッチするよう変更した。

ham-cap committed:
既存メモを取得する際の処理を変更

・hasOwnPropertyを使用して既存メモの存在確認をしていたが、getItemの返り値がnullか否かで判定でき、可読性も向上するため変更した。また、いずれの場合でもgetItemの実行回数を1回で済ませられるように処理の流れを変更した。

ham-cap committed:
aタグをbuttonタグに変更

・aタグで実装しているメモの一覧表示部分と新規登録フォームを呼び出す「+」は遷移先があるわけではなくhref属性はだが、href属性を付与しない場合、Tabフォーカスができなくなりマウスカーソルのホバー時も表示が変わらなくなる等の仕様があるため、buttonタグに変更しリンク風のスタイルを適用することとした。

ham-cap committed:
英語表記を日本語表記へ変更・フォームのボタン配置、間隔を調整

ham-cap committed:
Prettierによる修正

ham-cap committed:
closeCreateForm、closeEditFormの処理を変更

・フォームを閉じる前に既に何かが入力されていた場合、キャンセルしてもフォームにバインドされた各stateに値が残ってしまうため、キャンセル時には必ずstateを初期化するように変更した

yuuu commented on public/index.html:
```diff
@@ -14,8 +14,9 @@
     -->
-    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
     <!--
+    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
```
こちら、コメントアウトされているのはなぜでしょうか？アクセス元の端末ごとに適切なfaviconを返す、UIの設定をするという意味で必要なファイルだと認識しています。

ham-cap replied:
不要と思われるファイルや記述を削除した際、不要と勘違いして削除してしまいました💦
関連のファイルと合わせて復活させておきました🙇‍♂️

yuuu commented on src/App.jsx:
```diff
@@ -0,0 +1,138 @@
+    return (
+      <div>
+        <MemoList
+          memos={this.state.memos}
```
これだけ多くのpropsを渡さないといけない、ということはむしろ子コンポーネントで定義すべきハンドラやstateが混じっているのではないでしょうか？例えば `setSelectedMemo` `editingMemoText` 等は `App` ではなく `MemoList` 側に閉じた方が、使い回しがしやすくなりそうに見えます。

yuuu commented on src/FormSwitcher.jsx:
```diff
@@ -0,0 +1,49 @@
+
+export default class FormSwitcher extends React.Component {
+  render () {
+    let form
```
ここは `let` を使わずとも実現できるはずですので、そのように修正をお願いします 🙏

yuuu commented on src/MemoList.jsx:
```diff
@@ -0,0 +1,76 @@
+export default class MemoList extends React.Component {
+  render () {
+    const memos = this.props.memos
+    const memoList = memos.map((memo, index) => (
```
変数に入れたくなる、ということはここは子コンポーネントに切り出す余地があるということだと考えられます。 `Memo` コンポーネントに切り出してはどうでしょうか？

yuuu commented on src/MemoList.jsx:
```diff
@@ -0,0 +1,76 @@
+      </li>
+    ))
+
+    let form
```
ここも変数は不要ですね。
変数を使ってはいけないわけではないですが、変数を使うとDOMの構造を追いにくくなりますし、構造化したいのであればコンポーネントに切り出す方向で検討をお願い致します 🙏

ham-cap committed:
manifest.jsonの読み込みがコメントアウトされていたため復活させた

ham-cap committed:
CreateFormとEditFormの切り替え処理を修正

・letを使用して変数を宣言してからif文の返り値を代入していたが、関数の返り値を直接代入する形に変更した。

ham-cap committed:
Formの表示/非表示を切り替える部分を修正

・letを使用して変数を宣言してからif文の返り値を代入していたが、関数の返り値を直接代入する形に変更した。

ham-cap committed:
メモを一覧表示する際の処理を修正

・メモの一覧を表示する際、localStorageから取得したメモの配列を繰り返し処理でliタグのリストに変換したうえで変数に格納し、その変数をレンダリングしていた。
しかし、そもそもメモのリストを変数に入れる必要がなく、構造もわかりにくくなるため、変数に格納していた部分を丸ごとMemoコンポーネントとして切り出した。

ham-cap committed:
EditFormのレンダリング部分を修正

・変数formにEditFormを格納し、render内でformをレンダリングしていたが、変数に格納する必要がないため、returnの中でそのままisVisibleを判定してFormを返すように変更した。

ham-cap committed:
Appのstateや関数をMemoListに移した

・AppからMemoListに対して多くのpropsを渡していたが、それらがAppに定義されている必要がなかったため、MemoListに移した。

ham-cap committed:
MemoListが持っていたstateや関数をCreateFormへ移動

・MemoListから子コンポーネントへ渡すpropsが多すぎるため、MemoListが持つ必要性がないものを移動させた。

ham-cap committed:
Formの表示/非表示を切り替える部分を再度修正・デバッグ用の記述とpropsの不要なバリデーションを削除

・Formの表示を切り替える際、そもそも変数を用いる必要がなかったため修正した。

ham-cap commented:
ご指摘いただいた点を修正いたしましたのでご確認のほどよろしくお願いいたします🙏

yuuu commented on src/FormSwitcher.jsx:
```diff
@@ -0,0 +1,43 @@
+  render () {
+    if (this.props.isChanging) {
+      return (
+        <EditForm
```
`CreateForm` と `EditForm` を分けているのはなぜでしょうか？本来、CreateとEditの違いはタイトル部分とボタンの文言が変わるくらいの差分しかないはずです。

それ以外の点で処理の内容に差分が出ているのはなぜでしょうか 🤔

yuuu commented on src/EditForm.jsx:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  handleEditingMemoTextChange (e) {
+    this.props.onEditingMemoTextChange(e.target.value)
```
入力途中でこのコールバック関数を呼ぶ必要はあるのでしょうか？更新ボタンをクリックしたタイミングで `updateMemo` の引数で入力したメモの内容を渡す、という方法の方がシンプルではないでしょうか？

そうすれば `MemoList` の `editingMemoText` も削除できそうです。

yuuu commented on src/MemoList.jsx:
```diff
@@ -0,0 +1,133 @@
+    this.state = {
+      editingMemoText: '',
+      memos: [],
+      isCreating: false,
```
これら3つのフラグと `indexOfSelectedMemo` は全て必要でしょうか？
例えば `indexOfSelectedMemo` が `null` ではなければ、 `isChanging` であるといった判断は難しいでしょうか？

ham-cap committed:
CreateFormとEditFormを統合

・コンポーネントの構成が不必要に複雑になっていたため再検討したところ、CreateFormとEditFormに分割するのではなく、Formが表示された時のstateの状態によって、新規登録向けのフォームと更新用のフォームのいずれかが表示されるように変更した。

ham-cap committed:
使用しなくなったコンポーネントを削除・フォームのタイトルを変更

・CreateForm、EditForm、FormSwitcherは使用しなくなったため削除した
・フォームのタイトルが「Form」で固定されていたため、新規登録時と編集時にそれぞれ「新規登録」と「編集」と表示されるようにした

ham-cap commented:
いただいたヒントを参考に大幅に修正いたしました😃

CreateFormとEditFormはFormに統合しました。
また、コンポーネントの名称も実態に合わなくなってきたので変更し、結果としてAppの直下にMemo、その中にListとFormがあるという構造にしてみました。

> 入力途中でこのコールバック関数を呼ぶ必要はあるのでしょうか？更新ボタンをクリックしたタイミングで updateMemo の引数で入力したメモの内容を渡す、という方法の方がシンプルではないでしょうか？

この点についてなのですが、公式ドキュメントの[Controlled Component](https://ja.legacy.reactjs.org/docs/forms.html#controlled-components)で推奨されている方法を参考にしており、今回のプラクティスのヒントの欄でも
> Controlled Componentを使ってみよう

とあったのでこの方法で書いてみました！

上記のドキュメントを読んでみて、フォームからの入力はstateで保持し、それを更新できるのは`setState`のみという状態で運用するのがReactらしい書き方なのかなと解釈していたのですが、いかがでしょう？👀
理解が間違っていればご指摘いただけると幸いです🙏

yuuu commented on src/Form.jsx:
```diff
@@ -0,0 +1,60 @@
+  render () {
+    return (
+      <div>
+        {this.props.indexOfSelectedMemo ? <p>編集</p> : <p>新規登録</p>}
```
変えたいのは `<p>` タグの内側のテキストだけなので、そこだけを三項演算子にすると `<p>` タグの記述は1回で済みますね 👍

yuuu commented on src/Form.jsx:
```diff
@@ -0,0 +1,60 @@
+        />
+        {this.props.indexOfSelectedMemo
+          ? (
+          <div>
```
ここも一番外側の `<div>` タグは重複しているので三項演算子の外側に出しましょう。

yuuu replied:
@ham-cap 
こちらについて、外側を1要素にする必要があって意図的に `<div>` を残しているとのこと承知です 👍

yuuu commented on src/List.jsx:
```diff
@@ -0,0 +1,25 @@
+    return (
+      <ul>
+        {this.props.memos.map((memo, index) => (
+          <li key={index}>
```
以前コメントしたとおり、ここが1つのメモを表現している部分なので、ここが `Memo` コンポーネントと呼ぶに相応しいと思いました。コンポーネントに切り出さないとしても、今の `Memo` コンポーネントは別の名前を検討いただきたいです 🙏

yuuu commented on src/Form.jsx:
```diff
@@ -0,0 +1,60 @@
+        {this.props.indexOfSelectedMemo ? <p>編集</p> : <p>新規登録</p>}
+        <textarea
+          value={this.props.newMemoText}
+          onChange={this.props.handleNewMemoTextChange}
```
ControlledComponentsを参考にしていること自体は良いと思います 👍 ただ、公式ドキュメントに記載されているコンポーネントと、このコンポーネントでは2点違いがあると考えています。

1. Formコンポーネントの中でStateを持っていない
2. フォームに何か入力されるごとに `handleNewMemoTextChange` をコールバックしている

公式ドキュメントの `NameForm` はそれ単体で流用がしやすいと思います。一方、こちらの `Form` コンポーネントを流用しようとすると、呼び元で `textarea` の入力値を保持するStateを定義してもらう必要が生じてしまいますし、 `handleNewMemoTextChange` というコールバック関数を渡すことも強いています。これは `Form` を利用する立場からすると不便ではないでしょうか？

ただ、Qiitaのように「入力途中の内容も随時保存したいので `handleNewMemoTextChange` は必要」という可能性もあると思います。今回、それを意識してこのような設計にしているということであれば、`handleNewMemoTextChange` は残してもOKです。

ham-cap replied:
この点について煮詰まっているのでヒントをいただけないでしょうか👀

コメントでいただいた内容は、`Form`コンポーネントを利用する際に呼び出し元でしなくてはならないことが多すぎて使い勝手が悪いということが問題であり、`Form`コンポーネントをControlledComponentとして実装するということ自体に問題はないということだと理解しています。

そのうえで、フォームの表示/非表示の切り替えは親コンポーネントで行い、CRUDの処理は全て`Form`に下ろす方針で修正を進めています。
現状、新規登録については特に問題なく移行できたのですが、編集機能に伴って発生するフォームの初期値の扱いで困っています。

コンポーネントの構成は以下のように修正したうえで作業しています。
- App
    - MemosSection（←Memoコンポーネントの名称を変更しました）
        - List
            - Memo
        - Form

Formを編集用として表示する際、これまでは現在の`MemosSection`にあたるコンポーネント内で全てのstateを保持しており、`displayForm()`が`newMemoText`の更新まで一括して処理してくれていたおかげで、編集用のフォームを開きっぱなしの状態で別のメモをクリックしても、その都度`displayForm()`が`newMemoText`を更新してくれて、それを検知したReactがフォームを再レンダリングすることで初期値も自動で更新されるということが可能でした。

現在詰まっているのはこの点で、`Form`に`newMemoText`を下ろした状態で同じ動きを実現しようとすると、`displayForm()`から`newMemoText`を更新する処理を切り離すことになり、`displayForm()`が発火する（＝メモがクリックされる）タイミングで`newMemoText`を更新することができなくなります。
```js
  displayForm (e, index = null) {
    e.preventDefault()
    this.setState({
      isVisible: true,
      indexOfSelectedMemo: index
    })
// MemosSectionがnewMemoTextを持っていないとここから下の処理ができない。
    if (index !== null) {
      const selectedMemo = this.state.memos[index].join('\n')
      this.setSelectedMemo(selectedMemo)
    } else {
      this.setState({
        newMemoText: ''
      })
    }
  }
```
こうなると、一度編集フォームが表示されて初期値が設定された後は、別のメモをクリックしても`newMemoText`の更新処理が発火せず、フォームの初期値が最初にクリックしたもののままになってしまいます。
`Form`側で`componentDidMount`関数を利用して初期値をセットする方法などを試してみましたが、初回のマウント時にクリックされたメモの内容は初期値に入りますが、やはりその後は更新されません。
この場合、インデックスそのものは2回目以降にクリックされたメモのものに随時更新されており、意図しない対象が更新される可能性が出てきてしまうため、少なくとも現在の方針のまま進めるのであれば`newMemoText`の更新処理を`displayForm()`の発火タイミングと合わせることが必須になりそうだと考えているのですが、上手い方法が思いつかず、という状態です💦

同じ部分を考えすぎて視野が狭くなっているような気もするため、何かアドバイスをいただければ幸いです🙏

yuuu replied:
@ham-cap 
検討＆相談ありがとうございます 🙏 

> コメントでいただいた内容は、Formコンポーネントを利用する際に呼び出し元でしなくてはならないことが多すぎて使い勝手が悪いということが問題であり、FormコンポーネントをControlledComponentとして実装するということ自体に問題はないということだと理解しています。

こちらは認識あっています 👍 

> こうなると、一度編集フォームが表示されて初期値が設定された後は、別のメモをクリックしてもnewMemoTextの更新処理が発火せず、フォームの初期値が最初にクリックしたもののままになってしまいます。

ここについて、何となく状況は察しているのですが、現状のコードを見ながらアドバイスをしたいと考えています。お手数ですが動作しない状態でも構わないので、現状のコードを一度pushいただいても良いでしょうか？？

yuuu replied:
@ham-cap 
pushいただいたコードを拝見しました 👍 

### 起こっている問題

メモを編集する際、既にメモが選択されている状態で別のメモを選択しても、フォームの内容が切り替わらない。

→弊方でも確かにこの問題は起こっています。

### ham-capさんの見解

親コンポーネントである `MemosSection` にて `displayForm()` がコールバックされ `indexOfSelectedMemo` が変化するが、それを受け取った `Form` コンポーネント側で `newMemoText` の更新ができていない。

→弊方でも確認して、この見解は正しいと判断しています 👍 

### アドバイス

- `newMemoText` の更新処理は `componentDidMount` で行われていますが、 `componentDidMount` はどういったシチュエーションで呼ばれる関数か理解できてそうでしょうか？
    - 今回のように `props` の値が変化した場合、 `componentDidMount` は呼ばれる関数なのでしょうか？
    - ドキュメントの確認をしつつ、 `componentDidMount` の先頭に `console.log(this.props.indexOfSelectedMemo)` を入れて挙動も確認すると理解が深まると思います。
- もし、 `props` の更新時に `componentDidMount` が呼ばれない場合、 `props` の更新時に呼ばれる関数は他にないでしょうか？

ham-cap committed:
Listの中の単一のメモを表示する部分をMemoとして切り出した

・構造的にわかりやすいため単一のメモの部分を切り出した。それに伴って、既存のMemoコンポーネントの名称をMemosSectionへ変更した。

ham-cap committed:
フォームの表示切替時に三項演算子の中でdivタグが重複していた部分を外に出した

ham-cap committed:
新規登録と編集フォームを切り替える際の条件式を修正

・フォームの切り替えはindexOfSelectedMemoに値が入っているか否かで判定していたが、if文の条件式に0を記述した場合の真偽値はfalseが返ってしまい、結果としてインデックスが0の場合（＝localStorageに保存されているメモの中で最も古いものがクリックされた場合）にfalseとなり、既存のメモであるにもかかわらず新規登録フォームがレンダリングされてしまうため修正した。

ham-cap committed:
addMemoの引数にフォームの値を渡すよう変更

・これまではフォームに入力された値を親コンポーネントのnewMemoTextに同期させ、登録ボタンが押されたタイミングのnewMemoTextの値を保存する方法をとっていたが、コンポーネントの流用性が低下したり使い勝手が悪くなったりしてしまう。newMemoTextやhandleNewMemoTextChangeを廃止することでこれを改善するためにstateを利用せずフォームの値を直接addMemo関数へ渡すように変更した。

ham-cap committed:
メモの更新時、フォームの内容をstateではなくupdateMemoの引数として渡すようにした

ham-cap committed:
addMemoをFormへ移動

ham-cap committed:
updateMemoをFormへ移動

ham-cap commented:
取り急ぎ、現状のコードをpushいたしました🙏
ご確認よろしくお願いいたします🙇‍♂️

ham-cap committed:
componentDidUpdateを追加

・フォームを開きっぱなしにした状態で別のメモを選択してもフォームの初期値が最初に選択したメモのものから更新されなかったため修正した。

ham-cap committed:
deleteMemoをFormへ移動・不要なコメントアウト部分を削除

ham-cap committed:
MemosSectionのクラス名を変更・不要なstateを削除

・ファイル名の変更に合わせてクラス名を変更するのを失念していたため修正した。

ham-cap commented:
`componentDidUpdate`を使用して修正してみました😃
マウントとレンダリングを混同してしまっていたことと、最初に`componentDidUpdate`を使用したときに使い方の理解が足りておらずどうしても無限ループになってしまったことで解法の選択肢から外してしまっていました💧

また、以前[こちら](https://github.com/ham-cap/memo-app-react/pull/1#discussion_r1215414785)でご指摘いただいておりました`<div>`タグについてなのですが、条件が`true`の場合は`button`要素が2個になり、`<div>`を三項演算子の外に出すとLinterから`JSX expressions must have one parent element. `のエラーが出てしまうため、`false`に入った場合（新規登録用のフォームに入った場合）のみ`<div>`タグを削除し、`true`のほうはいったんそのままにしてあるのですが、解決法等ありますでしょうか？🤔

yuuu commented on src/Form.jsx:
```diff
@@ -0,0 +1,119 @@
+    ) {
+      this.setState({ newMemoText: '' })
+    } else if (
+      this.props.indexOfSelectedMemo !== prevProps.indexOfSelectedMemo
```
先に `this.props.indexOfSelectedMemo === prevProps.indexOfSelectedMemo` を満たす場合はアーリーリターンをしてしまって、次のif文で `this.props.indexOfSelectedMemo === null` を判定するとコードの重複がなくなってスッキリするはずです 👍

ham-cap replied:
この点について、例えば以下のようにアーリーリターンを試みるとLinterから`no-useless-return: Unnecessary return statement.`として怒られてしまいます。
関数を抜けるためだけに`return`を記述すること自体が推奨されていないということだと思うのですが、いただいたご指摘に対する私の理解がそもそもおかしかったりしますでしょうか？👀

```js
if (
  this.props.indexOfSelectedMemo === prevProps.indexOfSelectedMemo
) {
  return
} else if (
  this.props.indexOfSelectedMemo === null
) {
  this.setState({ newMemoText: '' })
} else {
  const selectedMemo =
    this.props.memos[this.props.indexOfSelectedMemo].join('\n')
  this.setState({ newMemoText: selectedMemo })
}
```

yuuu replied:
[こちらのドキュメント](https://eslint.org/docs/latest/rules/no-useless-return) を見る限り「必要ないreturnを書くべきではない」という指摘のようなので、アーリーリターンを試みると指摘が出る、というのは違うのでは？という気がしています。

例えば、

```javascript
const func = () => {
  if(/* 何らかの条件 */) {
    return;
  }
  /* 何らかの処理 */
}
```

のように書くべきところを

```javascript
const func = () => {
  if(/* 何らかの条件 */) {
    return;
  }
  /* 何らかの処理 */
  return;
}
```

と書いてしまっていて、「終わりの方の `return` は必要ないよね？」と指摘されている状況ではないでしょうか？

yuuu commented on src/MemosSection.jsx:
```diff
@@ -0,0 +1,65 @@
+    super(props)
+    this.state = {
+      memos: [],
+      isVisible: false,
```
`isVisible` と `indexOfSelectedMemo` は1つにまとめられそうです。 `indexOfSelectedMemo === null` であればCloseとみなせばよいので。2つに分けていることで片方だけ操作し忘れる、といったことを防げると思っています。

ham-cap replied:
`indexOfSelectedMemo === null`の場合にCloseとみなしてしまうと新規入力のフォームが開かなくなってしまいませんでしょうか？👀
`isVisible`はフォーム自体の表示/非表示、`indexOfSelectedMemo`はフォームが表示されるのは確定の状況で新規と編集どちらを表示するかを決定するために使用しており、別の役割を担っているため一つにまとめるイメージがつかないのですが🤔
もし理解が違っていたら教えてください🙇‍♂️

yuuu replied:
なるほど。新規入力の場合は `indexOfSelectedMemo === null` ということですね。ここは確かに見逃しており失礼いたしました 🙏 
このままでOKです 👍

yuuu commented on src/List.jsx:
```diff
@@ -0,0 +1,24 @@
+import PropTypes from 'prop-types'
+import Memo from './Memo.jsx'
+
+export default class List extends React.Component {
```
`List` という名前は今後リストで表現するものを扱う際に命名に困ってしまうので `MemoList` といった名前にしてはどうでしょう？

ham-cap committed:
componentDidUpdateを修正

・update前後のindexOfSelectedMemoを比較する部分でアーリーリターンを行うことでコードの重複をなくした。

ham-cap committed:
Listコンポーネントの名称をMemoListへ変更

・リストとして表現するものが増えた場合に判別できなくなるため、メモのリストであることを明示した。

ham-cap commented:
修正してみましたのでご確認をお願いいたします🙇‍♂️
